### PR TITLE
feature/refactoring

### PR DESCRIPTION
**Pull request by dovgopoly**

### CODE REVIEW

**Comment by Arvolear** [comment]
Please integrate the newest @dlsl/dev-modules 2.0.2

File: contracts/core/Swapica.sol [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+    modifier checkSignature(bytes calldata orderData, bytes[] calldata signatures) {
+        _checkSignatures(keccak256(orderData), signatures);
+        _;
+    }
+
+    function __Swapica_init(address[] calldata signers) external initializer {
+        __Signers_init(signers, signers.length);
+    }
+
+    function createOrder(CreateOrderRequest memory request) external payable override {
```
Comment (Arvolear) [inline contracts/core/Swapica.sol:37]:
```suggestion
    function createOrder(CreateOrderRequest calldata request) external payable override {
```

File: scripts/utils/utils.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import { ethers } from "hardhat";
+import { BigNumber } from "ethers";
+
+export function wei(value: string | number, decimal = 18) {
```
Comment (Arvolear) [inline scripts/utils/utils.ts:4]:
```suggestion
export function wei(value: string | number, decimal: number = 18) {
```

File: contracts/core/Swapica.sol [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    using Math for uint256;
+    using SafeERC20 for IERC20;
+
+    Order[] internal _orders;
+    Match[] internal _matches;
+
+    mapping(address => User) internal _userInfos;
+
+    event OrderUpdated(uint256 indexed orderId, OrderStatus status);
+    event MatchUpdated(uint256 indexed matchId, State status);
```
Comment (Arvolear) [inline contracts/core/Swapica.sol:26]: We should probably have 2 more events: OrderCreated, MatchCreated

File: test/utils/caster.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  if (!allKeys.some((key) => !Number.isInteger(+key)) || o instanceof BigNumber) {
+    return o;
+  }
+
+  return allKeys.reduce((prevKeys: any, currentKey: string) => {
+    if (Number.isInteger(+currentKey)) {
+      return prevKeys;
+    }
+
+    return { ...prevKeys, [currentKey]: castStruct(o[currentKey]) };
```
Comment (Arvolear) [inline test/utils/caster.ts:23]: The `[currentKey]` array construction looks like a crutch

File: contracts/core/Swapica.sol (hunk 1 above)
Comment (dovgopoly) [reply contracts/core/Swapica.sol]: done

File: scripts/utils/utils.ts (hunk 2 above)
Comment (dovgopoly) [reply scripts/utils/utils.ts]: done

File: contracts/core/Swapica.sol (hunk 3 above)
Comment (dovgopoly) [reply contracts/core/Swapica.sol:26]: done

File: contracts/core/Swapica.sol [hunk 5]
```diff
@@ -40,24 +42,23 @@ contract Swapica is ISwapica, UUPSUpgradeable, Signers {
...
+            amountToSell: request.amountToSell,
+            tokenToBuy: request.tokenToBuy,
+            amountToBuy: request.amountToBuy,
+            destinationChain: request.destinationChain
+        });
 
-        _userInfos[msg.sender].orderIds.push(orderId - 1);
+        _orders.push(order);
 
+        _userInfos[msg.sender].orderIds.push(orderId - 1);
```
Comment (Arvolear) [inline contracts/core/Swapica.sol:58]: Are you sure this `- 1` is needed?

Comment (dovgopoly) [reply contracts/core/Swapica.sol]: done

**Pull request merged**